packages/api/src/controllers/asset.ts

```diff
@@ -149,21 +193,31 @@ function parseUploadUrl(
 }
 
 app.use(
-  mung.json(function cleanWriteOnlyResponses(
+  mung.jsonAsync(async function cleanWriteOnlyResponses(
     data: WithID<Asset>[] | WithID<Asset> | { asset: WithID<Asset> },
     req
   ) {
-    if (req.user.admin) {
-      return data;
+    const ingests = await req.getIngest();
+    if (!ingests.length) {
+      throw new InternalServerError("Ingest not configured");
     }
+    const ingest = ingests[0].base;
+    const toExternalAsset = (a: WithID<Asset>) =>
+      req.user.admin
+        ? withPlaybackUrls(a, ingest)
+        : db.asset.cleanWriteOnlyResponse(withPlaybackUrls(a, ingest));
+
     if (Array.isArray(data)) {
```

Comment:
This section of the code is not clear to me, what are we doing here? In what case `data` has no `id`?

Reply:
This modifies every response from the controller to make sure we transform the output asset fields. This means basically clearing the fields that are private/admin-only when the calling user is not an admin as well as adding the playback URLs.

The response from the controller is not always a simple JSON object representing the asset. It can be an array of assets. It can also be an object with an asset object under the `asset` field. These are all the branches you see here. `id` field means the asset is the whole response object.